File: 프로그래머스/3/131703. 2차원 동전 뒤집기/2차원 동전 뒤집기.js
// 어떤 경우에 목표 상태를 만들 수 있고, 어떤 경우에 목표 상태를 만들 수 없을까?
// 행과 열을 뒤집는 순서는 상관 없다.
// post order traverse: 가능한 모든 경우 탐색하고 되돌려놓기
// space complexity: O(nm * 2^20)

const json = (value) => JSON.stringify(value);

function solution(beginning, target) {
    const current = beginning;
    
    let columnFlipCount = 0;
    let rowFlipCount = 0;
    for (let i = 0; i < target.length; i++) {
        let differenceCount = 0;
        for (let j = 0; j < target[i].length; j++) {
            if (i === 0 && current[i][j] !== target[i][j]) {
                for (let k = 0; k < target.length; k++) current[k][j] ^= 1;
                columnFlipCount += 1;
                continue;
            }
            
            if (current[i][j] !== target[i][j]) differenceCount += 1;
        }
        
        if (differenceCount !== 0 && differenceCount !== target[i].length) {
            return -1;
        } else if (differenceCount === target[i].length) rowFlipCount += 1;
    }
    
    return Math.min(columnFlipCount + rowFlipCount, 1 + (target[0].length - columnFlipCount) + (target.length - 1 - rowFlipCount));
    
}

File: 프로그래머스/3/131703. 2차원 동전 뒤집기/2차원 동전 뒤집기.py
from math import inf

def solution(beginning, target):
    result = dfs(0, 0, 0, beginning, target)
    if result == inf:
        return -1
    return result
    

def check_is_same(matrix, target):
    for i in range(0, len(matrix)):
        for j in range(0, len(matrix[i])):
            if matrix[i][j] != target[i][j]:
                return False
    
    return True

def dfs(row_count, col_count, switch_count, matrix, target):
    
    if (row_count == len(matrix) and col_count == len(matrix[0])):
        is_same = check_is_same(matrix, target)
        if is_same:
            return switch_count
        else:
            return inf
    
    min_count = inf
    
    if row_count < len(matrix):
        for i in range(0, len(matrix[0])):
            matrix[row_count][i] ^= 1
        min_count = min(min_count, dfs(row_count + 1, col_count, switch_count + 1, matrix, target))
        for i in range(0, len(matrix[0])):
            matrix[row_count][i] ^= 1
        
        min_count = min(min_count, dfs(row_count + 1, col_count, switch_count, matrix, target))
        return min_count
    else:
        for i in range(0, len(matrix)):
            matrix[i][col_count] ^= 1
        min_count = min(min_count, dfs(row_count, col_count + 1, switch_count + 1, matrix, target))
        for i in range(0, len(matrix)):
            matrix[i][col_count] ^= 1
            
        min_count = min(min_count, dfs(row_count, col_count + 1, switch_count, matrix, target))
        return min_count
    
    
    

    
    
    
            
        
        
    
    
                
    
                
    
                
    
    
    dfs(row_count,)
    

# 초기상태: 10111
# 목표상태: 11100